### ISSUE TITLE
cctbx_project/scitbx/suffixtree/test/tst_single.py segfaults with gcc 7.3

### ISSUE THREAD

**Issue by Anthchirp** [cctbx/cctbx_project#487]
Steps to reproduce starting with a RHEL 7.8 workstation and an empty working directory:
```bash
wget https://raw.githubusercontent.com/cctbx/cctbx_project/master/libtbx/auto_build/bootstrap.py
mv -n ~/.conda/environments.txt ~/.conda/environments.txt.preserved
python bootstrap.py --python 36 --use-conda hot update base --nproc=4
mkdir build; cd build
../conda_base/bin/python ../modules/cctbx_project/libtbx/configure.py  --use_conda libtbx scitbx
source setpaths.sh
```

<details><summary>Now compile with the system gcc (<tt>gcc --version</tt>)</summary>
<p>

```
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

</p>
</details>

```bash
libtbx.scons lib/scitbx_suffixtree_single_ext.so lib/scitbx_suffixtree_shared_ext.so lib/boost_python_meta_ext.so -j 4
```

at which point

<details><summary><tt>python ../modules/cctbx_project/scitbx/suffixtree/test/tst_single.py</tt> passes as expected</summary>
<p>

```
test_single_word (__main__.TestWord) ... ok
test_branch (__main__.TestEdge) ... ok
test_const_branch (__main__.TestEdge) ... ok
test_const_leaf (__main__.TestEdge) ... ok
test_const_root (__main__.TestEdge) ... ok
test_leaf (__main__.TestEdge) ... ok
test_root (__main__.TestEdge) ... ok
test_const_root (__main__.TestPreOrder) ... ok
test_r_b1 (__main__.TestPreOrder) ... ok
test_r_b1_b1 (__main__.TestPreOrder) ... ok
test_r_b2 (__main__.TestPreOrder) ... ok
test_root (__main__.TestPreOrder) ... ok
test_single_branch (__main__.TestPreOrder) ... ok
test_single_leaf (__main__.TestPreOrder) ... ok
test_single_root (__main__.TestPreOrder) ... ok
test_const_root (__main__.TestPostOrder) ... ok
test_r_b1 (__main__.TestPostOrder) ... ok
test_r_b1_b1 (__main__.TestPostOrder) ... ok
test_r_b2 (__main__.TestPostOrder) ... ok
test_root (__main__.TestPostOrder) ... ok
test_single_branch (__main__.TestPostOrder) ... ok
test_single_leaf (__main__.TestPostOrder) ... ok
test_single_root (__main__.TestPostOrder) ... ok
test_tree (__main__.TestTree) ... ok
test_ukkonen1 (__main__.TestTree) ... ok
test_ukkonen2 (__main__.TestTree) ... ok
test (__main__.TestMatchingStatistics) ... ok
test (__main__.TestLeafIndices) ... ok

----------------------------------------------------------------------
Ran 28 tests in 0.004s

OK
```

</p>
</details>

Now we install a newer compiler suite with
```bash
../mc3/bin/conda install -y -c conda-forge -p ../conda_base c-compiler cxx-compiler
export PATH=$(pwd)/../conda_base/bin:${PATH}
```
so that

<details><summary><tt>gcc --version</tt></summary>
<p>

```
gcc (crosstool-NG 1.23.0.449-a04d0) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

</p>
</details>

and then
```bash
rm -rf * .scon*
../conda_base/bin/python ../modules/cctbx_project/libtbx/configure.py  --use_conda libtbx scitbx
source setpaths.sh
libtbx.scons lib/scitbx_suffixtree_single_ext.so lib/scitbx_suffixtree_shared_ext.so lib/boost_python_meta_ext.so -j 4
```

finally

<details><summary><tt>python ../modules/cctbx_project/scitbx/suffixtree/test/tst_single.py</tt></summary>
<p>

```
test_single_word (__main__.TestWord) ... ok
test_branch (__main__.TestEdge) ... Segmentation fault (core dumped)
```

</p>
</details>
crashes with a segmentation fault.

We can now verify that it's **only** the compiler to blame with
```bash
../mc3/bin/conda remove -y -p ../conda_base c-compiler cxx-compiler
libtbx.scons lib/scitbx_suffixtree_single_ext.so lib/scitbx_suffixtree_shared_ext.so lib/boost_python_meta_ext.so -j 4
python ../modules/cctbx_project/scitbx/suffixtree/test/tst_single.py
```
and the test passes again.

**Comment by benjaminhwilliams** [comment]
I've just discovered that this problem, which seems to be causing build errors in DIALS, is affecting a downstream project, [xia2/screen19](https://github.com/xia2/screen19).  @bkpoon (or anyone else in the know), do you have a feeling of how much effort this would be to resolve?

**Comment by benjaminhwilliams** [comment]
I'm not a build manager of DIALS, so I don't know the details.  I just know that the continuous deployment is broken somehow and it's affecting me downstream.  I guess you'd have to look to someone in the know.  Perhaps @Anthchirp or @ndevenish?

**Comment by ndevenish** [comment]
Our internal deployment pipeline stops “stable nightly” releases on test errors, and the downstream job is dependent on that, so it’s not a build error, though affecting Ben internally to Diamond. We can probably do something about that separately.

I spent some time debugging this today and couldn’t get any more clarification. Without rebuilding, the source of the SEGV stays static but on a rebuild it moves, which sounds like classic corruption. I didn’t get to anything more forensic than that though, yet.

**Comment by bkpoon** [comment]
The test is failing for GCC > 5 and is expected to fail. The check for the gcc version only works for the compiler named gcc. I can add an additional check for the conda compilers until the test is fixed.

**Comment by graeme-winter** [comment]
@bkpoon if the test is expected to fail on modern compilers, maybe we should disable it or make it `xfail` or something for the moment while investigating?

**Comment by bkpoon** [comment]
@graeme-winter It is marked as expected to fail, but the check for the compiler version only checks for the compiler named `gcc`. Since the conda compilers have different names, the check is failing. I can update the check tomorrow.

**Comment by graeme-winter** [comment]
Ah, I did not connect the two clauses - OK.

**Comment by Anthchirp** [comment]
fixed by #489
